### ISSUE TITLE
Grafana Alert on Graph - InfluxDB error 400 if parameter contains backslash

### ISSUE THREAD

**Issue by DavidRawling** [grafana/grafana#10957]
Grafana Version: Grafana v4.6.3 (commit: 7a06a47) 
Datasource: InfluxDB 1.4.2-1 
OS: Debian 9.3 (Linux panel 4.9.0-5-amd64 #1 SMP Debian 4.9.65-3+deb9u2 (2018-01-04) x86_64 GNU/Linux)

This was also posted to the [Grafana community](https://community.grafana.com/t/debugging-an-alert-400-bad-request-data-graph-is-fine/5236/3) without response.

Current State:
* Dashboard configured with multiple Graph Panels.
* Problematic graph configured to retrieve % Disk Time and Queue Length for a disk volume with instance name "C:\Exchange\TransportDB" - this is a disk mounted on a path rather than having a drive letter. Query URL for Influx:

    http://influxhost:3000/api/datasources/proxy/1/query?db=telegraf&q=SELECT%20mean(%22Percent_Disk_Time%22)%20AS%20%22%25%20Disk%20Time%22%20FROM%20%22LogicalDisk%22%20WHERE%20(%22host%22%20%3D%20'EX'%20AND%20%22instance%22%20%3D%20'C%3A%5C%5CExchange%5C%5CTransportDB')%20AND%20time%20%3E%3D%20now()%20-%201h%20GROUP%20BY%20time(10s)%20fill(none)%3BSELECT%20mean(%22Avg._Disk_Queue_Length%22)%20AS%20%22Queue%20Length%22%20FROM%20%22LogicalDisk%22%20WHERE%20(%22host%22%20%3D%20'EX'%20AND%20%22instance%22%20%3D%20'C%3A%5C%5CExchange%5C%5CTransportDB')%20AND%20time%20%3E%3D%20now()%20-%201h%20GROUP%20BY%20time(10s)%20fill(none)&epoch=ms

Decoded with line-breaks for legibility:

    http://influxhost:3000/api/datasources/proxy/1/query?db=telegraf&q=
        SELECT mean("Percent_Disk_Time") AS "% Disk Time"
        FROM "LogicalDisk" 
        WHERE ("host" = 'EX' AND "instance" = 'C:\\Exchange\\TransportDB') AND time >= now() - 1h
        GROUP BY time(10s) fill(none);
        SELECT mean("Avg._Disk_Queue_Length") AS "Queue Length" 
        FROM "LogicalDisk" 
        WHERE ("host" = 'EX' AND "instance" = 'C:\\Exchange\\TransportDB') AND time >= now() - 1h
        GROUP BY time(10s) fill(none)&epoch=ms

* Graph displays the correct data (e.g. % Disk Time = 0%, Queue Length = 0.423)

If I now configure an alert for that datum, the alert will throw error 400:

* Conditions:
    * WHEN avg() OF query (A, 5m, now) IS ABOVE 95
    * AND avg() OF query (B, 5m, now) IS ABOVE 16
* Error from Test Rule:
    * firing:true
    * state:""
    * conditionEvals:" = true"
    * timeMs:"1.053ms"
    * error:"tsdb.HandleRequest() error Influxdb returned statuscode invalid status code: 400 Bad Request"

I can't see a way to get the alert query and response - grafana.log displays only the details below, and it doesn't appear in the browser debugger:

t=2018-02-18T10:58:46+0000 lvl=eror msg="Alert Rule Result Error" logger=alerting.evalHandler ruleId=40 name="EX - Hard Disk Performance - Transport" error="tsdb.HandleRequest() error Influxdb returned statuscode invalid status code: 400 Bad Request" changing state to=keep_state

Panel JSON (compacted):

`{"alert":{"conditions":[{"evaluator":{"params":[95],"type":"gt"},"operator":{"type":"and"},"query":{"params":["A","5m","now"]},"reducer":{"params":[],"type":"avg"},"type":"query"},{"evaluator":{"params":[16],"type":"gt"},"operator":{"type":"and"},"query":{"params":["B","5m","now"]},"reducer":{"params":[],"type":"avg"},"type":"query"}],"executionErrorState":"keep_state","frequency":"60s","handler":1,"name":"EX - Hard Disk Performance - Transport","noDataState":"no_data","notifications":[{"id":2},{"id":3}]},"aliasColors":{},"bars":false,"dashLength":10,"dashes":false,"datasource":null,"fill":1,"id":22,"legend":{"avg":false,"current":false,"max":false,"min":false,"show":true,"total":false,"values":false},"lines":true,"linewidth":1,"links":[{"type":"dashboard","dashboard":"EX","dashUri":"db/ex","title":"EX"}],"nullPointMode":"null","percentage":false,"pointradius":5,"points":false,"renderer":"flot","seriesOverrides":[{"alias":"Queue Length","yaxis":2}],"spaceLength":10,"span":3,"stack":false,"steppedLine":false,"targets":[{"alias":"$col","dsType":"influxdb","groupBy":[{"params":["$__interval"],"type":"time"},{"params":["none"],"type":"fill"}],"measurement":"LogicalDisk","orderByTime":"ASC","policy":"default","refId":"A","resultFormat":"time_series","select":[[{"params":["Percent_Disk_Time"],"type":"field"},{"params":[],"type":"mean"},{"params":["% Disk Time"],"type":"alias"}]],"tags":[{"key":"host","operator":"=","value":"EX"},{"condition":"AND","key":"instance","operator":"=","value":"C:\\Exchange\\TransportDB"}]},{"alias":"$col","dsType":"influxdb","groupBy":[{"params":["$__interval"],"type":"time"},{"params":["none"],"type":"fill"}],"measurement":"LogicalDisk","orderByTime":"ASC","policy":"default","refId":"B","resultFormat":"time_series","select":[[{"params":["Avg._Disk_Queue_Length"],"type":"field"},{"params":[],"type":"mean"},{"params":["Queue Length"],"type":"alias"}]],"tags":[{"key":"host","operator":"=","value":"EX"},{"condition":"AND","key":"instance","operator":"=","value":"C:\\Exchange\\TransportDB"}]}],"thresholds":[{"value":95,"op":"gt","fill":true,"line":true,"colorMode":"critical"}],"timeFrom":null,"timeShift":null,"title":"EX - Hard Disk Performance - Transport","tooltip":{"shared":true,"sort":0,"value_type":"individual"},"type":"graph","xaxis":{"buckets":null,"mode":"time","name":null,"show":true,"values":[]},"yaxes":[{"decimals":null,"format":"percent","label":null,"logBase":1,"max":"100","min":"0","show":true},{"format":"short","label":null,"logBase":1,"max":null,"min":null,"show":true}]}`

It is possible this is the same as #6488 and its relatives.

Expected behaviour: Alert configured and correctly evaluated. I have another graph and alert set with configuration identical apart from the disk label (C: instead of C:\Exchange\TransportDB) - this one works correctly. Cloning and updating the working graph produces the same behaviour as configuring from scratch.

**Comment by daniellee** [comment]
Just answered on the community site and then saw you had posted here.

It is a possible bug. The backslash is escaped on the frontend as a double backslash. Checking the backend to see if it escapes backslashes properly.

**Comment by daniellee** [comment]
Testing with the latest version of Grafana I'm not getting the same error but the escaping logic for backslashes does not seem to be implemented on the backend.

This is the [code for rendering tags on the backend](https://github.com/grafana/grafana/blob/master/pkg/tsdb/influxdb/query.go#L67-L74):

```js
// quote value unless regex or number
		if tag.Operator == "=~" || tag.Operator == "!~" {
			textValue = tag.Value
		} else if tag.Operator == "<" || tag.Operator == ">" {
			textValue = tag.Value
		} else {
			textValue = fmt.Sprintf("'%s'", tag.Value)
		}
```

This is the [code for rendering tags on the frontend](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/influxdb/influx_query.ts#L143-L153):

```js
// quote value unless regex
    if (operator !== '=~' && operator !== '!~') {
      if (interpolate) {
        value = this.templateSrv.replace(value, this.scopedVars);
      }
      if (operator !== '>' && operator !== '<') {
        value = "'" + value.replace(/\\/g, '\\\\') + "'";
      }
    } else if (interpolate) {
      value = this.templateSrv.replace(value, this.scopedVars, 'regex');
    }
```

@bergquist looks like the backslash replace code is missing on the backend or am I missing something?

**Comment by bergquist** [comment]
Looks like it. Can't think of any good reason for not doing it. 
Argh. Frustrating.

**Comment by bergquist** [comment]
@daniellee you assigned this to yourself. Do you want to fix it or should I do it?

**Comment by DavidRawling** [comment]
Thank you both. Is there an easy way to tell which build(s) of Grafana will include the fix - will it be only 5.0.1 or later, given how close 5.0.0 seems to be to release?

**Comment by bergquist** [comment]
@DavidRawling since its such a small fix it will be included in 5.0
